cargo-guppy/src/lib.rs

```diff
@@ -18,18 +25,136 @@ pub fn cmd_diff(json: bool, old: &str, new: &str) -> Result<(), Error> {
     Ok(())
 }
 
-pub fn cmd_count() -> Result<(), Error> {
+pub fn cmd_count() -> Result<(), anyhow::Error> {
     let lockfile = Lockfile::from_file("Cargo.lock")?;
 
     println!("Third-party Packages: {}", lockfile.third_party_packages());
 
     Ok(())
 }
 
-pub fn cmd_dups() -> Result<(), Error> {
+pub fn cmd_dups() -> Result<(), anyhow::Error> {
     let lockfile = Lockfile::from_file("Cargo.lock")?;
 
     lockfile.duplicate_packages();
 
     Ok(())
 }
+
+arg_enum! {
+    #[derive(Debug)]
+    pub enum Kind {
+        All,
+        Workspace,
+        ThirdParty,
+    }
+}
+
+struct NameVisitor;
+
+impl PackageDotVisitor for NameVisitor {
+    fn visit_package(&self, package: &PackageMetadata, mut f: DotWrite<'_, '_>) -> fmt::Result {
+        write!(f, "{}", package.name())
+    }
+
+    fn visit_link(&self, _link: DependencyLink<'_>, mut f: DotWrite<'_, '_>) -> fmt::Result {
+        write!(f, "")
+    }
+}
+
+pub fn cmd_select(
+    kind: Kind,
+    target: Option<String>,
+    output: Option<String>,
+    include_dev: bool,
+    include_build: bool,
+    omit_edges_into: &[String],
+    roots: &[String],
+) -> Result<(), anyhow::Error> {
+    let mut pkg_graph = PackageGraph::default();
+    let mut package_ids = HashSet::new();
+    let mut omitted_package_ids = HashSet::new();
+
+    let root_set: HashSet<String> = if roots.len() > 0 {
+        roots.iter().cloned().collect()
+    } else {
+        pkg_graph
+            .packages()
+            .filter_map(|meta| {
+                if meta.in_workspace() {
+                    Some(meta.name().to_string())
+                } else {
+                    None
+                }
+            })
+            .collect()
+    };
+
+    let omitted_set: HashSet<String> = omit_edges_into.iter().cloned().collect();
+
+    for metadata in pkg_graph.packages() {
+        if root_set.contains(metadata.name()) {
+            package_ids.insert(metadata.id().clone());
+        }
+        if omitted_set.contains(metadata.name()) {
+            omitted_package_ids.insert(metadata.id().clone());
+        }
+    }
+
+    let transitive_deps: Vec<_> = pkg_graph
+        .select_forward(&package_ids)?
+        .into_iter_ids(None)
+        .chain(&package_ids)
```

Comment:
Hmm I don't think this `chain` is necessary since transitive deps includes the package ids themselves. Please let me know if the documentation needs to be better somehow?

Reply:
Ah, I didn't notice this from your new changes, or else I had it wrong all along.